icechunk/src/stream_utils.rs

```diff
@@ -0,0 +1,123 @@
+use std::{collections::HashSet, marker::PhantomData};
+
+use futures::{TryStream, TryStreamExt};
+use tokio::sync::{Mutex, Notify};
+use tracing::trace;
+
+/// Return the unique elements of the input stream
+/// Unique is computed according to `f`, and only the first
+/// matching element is yielded.
+pub(crate) fn try_unique_stream<S, T, E, F, V>(
+    f: F,
+    stream: S,
+) -> impl TryStream<Ok = T, Error = E>
+where
+    F: Fn(&S::Ok) -> V,
+    S: TryStream<Ok = T, Error = E>,
+    V: Eq + std::hash::Hash,
+{
+    let mut seen = HashSet::new();
+    stream.try_filter(move |item| {
+        let v = f(item);
+        if seen.insert(v) {
+            futures::future::ready(true)
+        } else {
+            futures::future::ready(false)
+        }
+    })
+}
+
+#[derive(Debug, Default)]
+struct StreamLimiterState {
+    // TODO: we may want to switch to atomic values instead
+    current_usage: usize,
+    in_flight: usize,
+}
+
+/// A component to limit the rate at which a stream produces items
+/// Call `limit` on each element of the stream before they are processed.
+/// Call `free` on each element of the stream once they are done.
+/// No more than `max_usage` usage will be generated in total by the in-flight
+/// yielded elements. Usage is measured according to the function passed to `new`.
+/// Usually, this is used in combination with `TryStreamExt::try_buffer`.
+///
+/// Note: this could have a much better interface with a bit of extra work:
+/// - We could simple wrap the stream and it would take care of mapping every element
+/// - Elements could be freed automatically using the Drop instance
+#[derive(Debug)]
+pub(crate) struct StreamLimiter<T, F: Fn(&'_ T) -> usize> {
+    get_usage: F,
+    max_usage: usize,
+    state: Mutex<StreamLimiterState>,
+    waiter: Notify,
+    _phantom: PhantomData<T>,
```

Comment:
Why do we need the phantom?

Reply:
Rust doesn't like a type declared as `Foo<T>` not to use `T`. `PhantomData` has the mission of making the compiler think I'm using it, and still not take any storage.